### PR TITLE
network/socket: socket binding on TCB

### PR DESCRIPTION
**Pull request by deepaksrma**
- with this patch, sockets binds with TCB
- when task kills then it releases its socket resources

Signed-off-by: Deepak sharma <deepak.sh@samsung.com>

### CODE REVIEW

File: os/net/lwip/src/api/sockets.c [hunk 1]
```diff
@@ -453,28 +468,31 @@ struct lwip_sock *trycheck_selwait_socket(int s)
...
+
+	if (!sock) {
+		ndbg("memory is not sufficient for sockets\n");
+		return -1;
+	}
 
-	/* allocate a new socket identifier */
-	for (i = 0; i < NUM_SOCKETS; ++i) {
-		/* Protect socket array */
+	if (list) {
```
Comment (sunghan-chang) [inline os/net/lwip/src/api/sockets.c:484]: Don't you need to add else case to free the sock for error case?

File: os/net/lwip/src/api/sockets.c [hunk 2]
```diff
@@ -299,13 +299,11 @@ struct lwip_socket_multicast_pair {
 
 struct lwip_socket_multicast_pair socket_ipv4_multicast_memberships[LWIP_SOCKET_MAX_MEMBERSHIPS];
 
-static int lwip_socket_register_membership(int s, const ip4_addr_t *if_addr, const ip4_addr_t *multi_addr);
-static void lwip_socket_unregister_membership(int s, const ip4_addr_t *if_addr, const ip4_addr_t *multi_addr);
-static void lwip_socket_drop_registered_memberships(int s);
+static int lwip_socket_register_membership(int s, const ip4_addr_t *if_addr, const ip4_addr_t *multi_addr, pid_t);
+static void lwip_socket_unregister_membership(int s, const ip4_addr_t *if_addr, const ip4_addr_t *multi_addr, pid_t pid);
+static void lwip_socket_drop_registered_memberships(struct lwip_sock *);
```
Comment (sunghan-chang) [inline os/net/lwip/src/api/sockets.c:304]: You should add the argument name.
`struct lwip_sock *` is not allowed.

File: os/net/lwip/src/api/sockets.c [hunk 3]
```diff
@@ -1498,7 +1529,7 @@ int lwip_select(int maxfdp1, fd_set *readset, fd_set *writeset, fd_set *exceptse
 			if ((readset && FD_ISSET(i, readset)) || (writeset && FD_ISSET(i, writeset)) || (exceptset && FD_ISSET(i, exceptset))) {
 				struct lwip_sock *sock;
 				SYS_ARCH_PROTECT(lev);
-				sock = tryget_socket(i);
+				sock = tryget_socket(i, pid);
```
Comment (sunghan-chang) [inline os/net/lwip/src/api/sockets.c:1532]: I can't find a local variable for pid. Is that a global variable?

Comment (deepaksrma) [reply os/net/lwip/src/api/sockets.c]: thanks and corrected now.

File: os/net/lwip/src/api/sockets.c (hunk 2 above)
Comment (deepaksrma) [reply os/net/lwip/src/api/sockets.c]: done

File: os/net/lwip/src/api/sockets.c (hunk 1 above)
Comment (deepaksrma) [reply os/net/lwip/src/api/sockets.c]: thanks and corrected.

File: os/net/lwip/src/api/sockets.c [hunk 4]
```diff
@@ -681,6 +696,22 @@ int lwip_sock_close(struct lwip_sock *sock)
 {
 	err_t err;
 	int is_tcp = 0;
+	struct socketlist list;
+	FAR struct tcb_s *tcb;
+	int sd;
+	pid_t pid = sock->conn->pid;
+
+	if (!sock) {
+		return -1;
```
[the remark this answers is not in the document]
Comment (deepaksrma) [reply os/net/lwip/src/api/sockets.c:706]: corrected.

File: os/net/lwip/src/api/sockets.c [hunk 5]
```diff
@@ -1265,8 +1297,8 @@ int lwip_socket(int domain, int type, int protocol)
 		set_errno(ENOBUFS);
 		return -1;
 	}
-
 	i = alloc_socket(conn, 0);
+	ndbg("socket = %d, pid = %d, %s: %s: %d\n", i, getpid(), __FUNCTION__, __FILE__, __LINE__);
```
[the remark this answers is not in the document]
Comment (deepaksrma) [reply os/net/lwip/src/api/sockets.c]: done.

File: os/net/lwip/src/api/sockets.c [hunk 6]
```diff
@@ -1265,8 +1299,8 @@ int lwip_socket(int domain, int type, int protocol)
 		set_errno(ENOBUFS);
 		return -1;
 	}
-
 	i = alloc_socket(conn, 0);
+	nvdbg("socket = %d, pid = %d, %s: %s: %d\n", i, getpid(), __FILE__, __LINE__);
```
Comment (sunghan-chang) [inline os/net/lwip/src/api/sockets.c:1303]: Are there many `lwip_socket` functions? I don't know why you need __FILE__ and __LINE__. If there is only one, I think they are unnecessary.

File: os/net/netmgr/netstack_lwip.c [hunk 7]
```diff
@@ -474,6 +475,34 @@ static int lwip_ns_getstats(int fd, struct netmon_sock **sock_info)
...
+	int i;
+	int ret;
+	struct lwip_sock *psock;
+	for (i = 0; i < CONFIG_NSOCKET_DESCRIPTORS; ++i) {
+		if (list->sl_sockets[i].sock) {
+			psock = (struct lwip_sock *)list->sl_sockets[i].sock;
+			DEBUGASSERT(psock->conn);
+			ret = lwip_sock_close(psock);
+			if (ret) {
+				ndbg("socket could not close properly\n");
```
Comment (sunghan-chang) [inline os/net/netmgr/netstack_lwip.c:499]: There is no exit with error case. Is it ok to set it as NULL? It will be memory leak.

File: os/net/socket/net_sockets.c [hunk 8]
```diff
@@ -133,10 +136,22 @@ void net_initlist(FAR struct socketlist *list)
...
+	int i;
+	int ret;
+	struct lwip_sock *psock;
+	for (i = 0; i < CONFIG_NSOCKET_DESCRIPTORS; ++i) {
+		if (list->sl_sockets[i].sock) {
+			psock = (struct lwip_sock *)list->sl_sockets[i].sock;
+			DEBUGASSERT(psock->conn);
+			ret = lwip_sock_close(psock);
+			if (ret) {
+				ndbg("socket could not close properly\n");
```
Comment (sunghan-chang) [inline os/net/socket/net_sockets.c:148]: same as previous

File: os/net/lwip/src/api/sockets.c [hunk 9]
```diff
@@ -2947,9 +2980,9 @@ static int lwip_socket_register_membership(int s, const ip4_addr_t *if_addr, con
  *
  * ATTENTION: this function is called from tcpip_thread (or under CORE_LOCK).
  */
-static void lwip_socket_unregister_membership(int s, const ip4_addr_t *if_addr, const ip4_addr_t *multi_addr)
+static void lwip_socket_unregister_membership(int s, const ip4_addr_t *if_addr, const ip4_addr_t *multi_addr, pid_t pid)
```
Comment (sunghan-chang) [inline os/net/lwip/src/api/sockets.c:2983]: `pid` is not used.
I have queries for that.
1. Is it ok to treat (or handle) other's socket by pid value?
2. If 1st is wrong, how about moving `getpid` into `get_socket` inside?

**Comment by sunghan-chang** [comment]
@deepaksrma Could you confirm this is verified on both mtk and rtk?

File: os/net/lwip/src/api/sockets.c [hunk 10]
```diff
@@ -1454,7 +1487,7 @@ int lwip_select(int maxfdp1, fd_set *readset, fd_set *writeset, fd_set *exceptse
 			if ((readset && FD_ISSET(i, readset)) || (writeset && FD_ISSET(i, writeset)) || (exceptset && FD_ISSET(i, exceptset))) {
 				struct lwip_sock *sock;
 				SYS_ARCH_PROTECT(lev);
-				sock = tryget_socket(i);
+				sock = tryget_socket(i, getpid());
```
Comment (jeongchanKim) [inline os/net/lwip/src/api/sockets.c:1490]: There are many `getpid()` in this function. How about declaring one local variable and use that if there is no context switching between every `getpid()`?

File: os/net/lwip/src/api/sockets.c [hunk 11]
```diff
@@ -1817,8 +1849,8 @@ static void event_callback(struct netconn *conn, enum netconn_evt evt, u16_t len
 			s = conn->socket;
 			SYS_ARCH_UNPROTECT(lev);
 		}
-
-		sock = get_socket(s);
+		pid = conn->pid;
+		sock = get_socket(s, pid);
```
Comment (jeongchanKim) [inline os/net/lwip/src/api/sockets.c:1853]: I think we don't need to use local variable, because `conn->pid` is only used once.

File: os/net/netmgr/netstack_lwip.c [hunk 12]
```diff
@@ -427,22 +427,23 @@ static void _get_raw_info(struct netmon_sock *sock_info, struct lwip_sock *lsock
...
  */
-extern struct lwip_sock *get_lwip_sock_info(void);
+extern struct socketlist *get_lwip_sock_info(void);
 
 static int lwip_ns_getstats(int fd, struct netmon_sock **sock_info)
 {
-	struct lwip_sock *sockets = get_lwip_sock_info();
-	if (!sockets[fd].conn) {
+	struct socketlist *list = get_lwip_sock_info();
+	struct lwip_sock *sock = (struct lwip_sock *) list->sl_sockets[fd].sock;
```
Comment (jeongchanKim) [inline os/net/netmgr/netstack_lwip.c:442]: You'd better check `list` is NULL or not. Sometimes it can be dereferencing for `list->sl_sockets`.

File: os/net/socket/net_sockets.c [hunk 13]
```diff
@@ -133,10 +136,22 @@ void net_initlist(FAR struct socketlist *list)
...
 void net_releaselist(FAR struct socketlist *list)
 {
-	DEBUGASSERT(list);
+	int i;
+	int ret;
+	struct lwip_sock *psock;
+	for (i = 0; i < CONFIG_NSOCKET_DESCRIPTORS; ++i) {
+		if (list->sl_sockets[i].sock) {
+			psock = (struct lwip_sock *)list->sl_sockets[i].sock;
+			DEBUGASSERT(psock->conn);
```
Comment (jeongchanKim) [inline os/net/socket/net_sockets.c:145]: Why did you check `psock->conn` is NULL or not? `psock->conn` is not used anywhere in this function.

File: os/net/lwip/src/api/sockets.c [hunk 14]
```diff
@@ -453,31 +468,36 @@ struct lwip_sock *trycheck_selwait_socket(int s)
...
 {
-	int i;
+	int idx;
+	struct socketlist *list;
+	struct lwip_sock *sock;
+
+	list = sched_getsockets();
+	if (!list) {
+		ndbg("could not get the socketlist\n");
+		return -1;
```
Comment (amandeepc) [inline os/net/lwip/src/api/sockets.c:478]: maybe can use ERROR in place of -1

File: os/net/lwip/src/api/sockets.c [hunk 15]
```diff
@@ -560,7 +580,8 @@ int lwip_accept(int s, struct sockaddr *addr, socklen_t *addrlen)
 	}
 	LWIP_ASSERT("invalid socket index", (newsock >= LWIP_SOCKET_OFFSET) && (newsock < NUM_SOCKETS + LWIP_SOCKET_OFFSET));
 	LWIP_ASSERT("newconn->callback == event_callback", newconn->callback == event_callback);
-	nsock = &sockets[newsock - LWIP_SOCKET_OFFSET];
+	//nsock = &sockets[newsock - LWIP_SOCKET_OFFSET];
```
Comment (amandeepc) [inline os/net/lwip/src/api/sockets.c:583]: can remove this comment

File: os/net/lwip/src/api/sockets.c (hunk 14 above)
Comment (amandeepc) [reply os/net/lwip/src/api/sockets.c:478]: or something similar

File: os/net/netmgr/netstack_lwip.c (hunk 12 above)
Comment (amandeepc) [inline os/net/netmgr/netstack_lwip.c:442]: no space here i believe b/w typecasting and list->sl_sockets[fd].sock

File: os/net/lwip/src/api/sockets.c (hunk 6 above)
Comment (deepaksrma) [reply os/net/lwip/src/api/sockets.c]: corrected.

File: os/net/netmgr/netstack_lwip.c (hunk 7 above)
Comment (deepaksrma) [reply os/net/netmgr/netstack_lwip.c:500]: corrected

File: os/net/socket/net_sockets.c (hunk 8 above)
Comment (deepaksrma) [reply os/net/socket/net_sockets.c:149]: corrected

File: os/net/lwip/src/api/sockets.c (hunk 9 above)
Comment (deepaksrma) [reply os/net/lwip/src/api/sockets.c:2977]: pid is used now.

File: os/net/lwip/src/api/sockets.c (hunk 11 above)
Comment (deepaksrma) [reply os/net/lwip/src/api/sockets.c]: modified

File: os/net/socket/net_sockets.c (hunk 13 above)
Comment (deepaksrma) [reply os/net/socket/net_sockets.c:146]: because if socket is there but connection is not there then this is a problem.

File: os/net/lwip/src/api/sockets.c (hunk 14 above)
Comment (deepaksrma) [reply os/net/lwip/src/api/sockets.c:478]: when socket is not created then earlier they were returning -1 so i thought let's keep it same. seems no problem here.

File: os/net/lwip/src/api/sockets.c (hunk 10 above)
Comment (deepaksrma) [reply os/net/lwip/src/api/sockets.c:1486]: this looks more simple than your suggested method. because according to you method each time we have to compare local var with getpid() to check whether pid is changed or not. so i found it more simpler.

File: os/net/netmgr/netstack_lwip.c (hunk 12 above)
Comment (deepaksrma) [reply os/net/netmgr/netstack_lwip.c]: done

File: os/net/lwip/src/api/sockets.c (hunk 15 above)
Comment (deepaksrma) [reply os/net/lwip/src/api/sockets.c]: done

File: os/net/netmgr/netstack_lwip.c (hunk 12 above)
Comment (deepaksrma) [reply os/net/netmgr/netstack_lwip.c]: kept in mind. next time i'll change it.

File: os/net/socket/net_sockets.c (hunk 13 above)
Comment (jeongchanKim) [reply os/net/socket/net_sockets.c:146]: In that case, I think we'd better check it when it is used.

**Comment by deepaksrma** [comment]
> @deepaksrma Could you confirm this is verified on both mtk and rtk?

Ahreum and i verified on the mtk board but not on rtk.
will verify on rtk also.

File: os/include/tinyara/net/net.h [hunk 16]
```diff
@@ -174,6 +174,7 @@ struct socket {
 	/** sockets currently are built on netconns, each socket has one netconn */
 	void *conn;
 	sock_type type; // lwip_sock, uds, ...
+	void *sock;
```
Comment (kishore-sn) [inline os/include/tinyara/net/net.h:177]: Please add a comment to explain the meaning of this variable.

File: os/net/lwip/src/api/sockets.c [hunk 17]
```diff
@@ -367,29 +365,35 @@ void lwip_socket_thread_cleanup(void)
...
 		set_errno(EBADF);
 		return NULL;
 	}
 
-	sock = &sockets[s];
+	tcb = sched_gettcb(pid);
+	DEBUGASSERT(tcb);
 
-	if (!sock->conn) {
+	list = tcb->group->tg_socketlist;
```
Comment (kishore-sn) [inline os/net/lwip/src/api/sockets.c:393]: Pls check for validity of tcb-> group before using it.
DEBUGASSERT(tcb && tcb->group);

File: os/net/lwip/src/api/sockets.c [hunk 18]
```diff
@@ -453,31 +468,36 @@ struct lwip_sock *trycheck_selwait_socket(int s)
...
 int alloc_socket(struct netconn *newconn, int accepted)
 {
-	int i;
+	int idx;
+	struct socketlist *list;
+	struct lwip_sock *sock;
+
+	list = sched_getsockets();
+	if (!list) {
+		ndbg("could not get the socketlist\n");
```
Comment (kishore-sn) [inline os/net/lwip/src/api/sockets.c:477]: It looks like LWIP_DEBUGF is used in most places in this file. Why are you using ndbg? Can we make it uniform?

File: os/net/lwip/src/api/sockets.c [hunk 19]
```diff
@@ -399,27 +403,41 @@ struct lwip_sock *get_socket(int s)
...
+	FAR struct tcb_s *tcb;
 	s -= LWIP_SOCKET_OFFSET;
-	if ((s < 0) || (s >= NUM_SOCKETS)) {
+
+	if ((s < 0) || (s >= CONFIG_NSOCKET_DESCRIPTORS)) {
 		return NULL;
 	}
-	if (!sockets[s].conn) {
+
+	tcb = sched_gettcb(pid);
```
Comment (jeongarmy) [inline os/net/lwip/src/api/sockets.c:431]: This is same with sched_getsockets()

File: os/net/lwip/src/api/sockets.c [hunk 20]
```diff
@@ -453,31 +468,36 @@ struct lwip_sock *trycheck_selwait_socket(int s)
...
+	if (!list) {
+		ndbg("could not get the socketlist\n");
+		return -1;
+	}
+
 	SYS_ARCH_DECL_PROTECT(lev);
+	sock = (struct lwip_sock *)kmm_zalloc(sizeof(struct lwip_sock));
+	if (!sock) {
+		ndbg("memory is not sufficient for sockets\n");
+		return -1;
```
Comment (jeongarmy) [inline os/net/lwip/src/api/sockets.c:485]: please set ERROR or ENOMEM instead of -1

File: os/net/socket/net_sockets.c (hunk 13 above)
Comment (deepaksrma) [reply os/net/socket/net_sockets.c:146]: I didn't get your point clearly. Here, i am checking socket conn because i am fetching socket descriptor from netconn and with that SD i am able assign NULL to corresponding socket after releasing the socket resources.

File: os/net/lwip/src/api/sockets.c (hunk 19 above)
Comment (deepaksrma) [reply os/net/lwip/src/api/sockets.c:431]: how it is same with sched_getsockets()?  because i am fetching corresponding tcb from the pid.

File: os/net/lwip/src/api/sockets.c (hunk 20 above)
Comment (deepaksrma) [reply os/net/lwip/src/api/sockets.c:485]: it will not matter much because when it will return -1, there it set ENFILE as error.

File: os/net/lwip/src/api/sockets.c (hunk 18 above)
Comment (deepaksrma) [reply os/net/lwip/src/api/sockets.c]: done

File: os/net/lwip/src/api/sockets.c (hunk 17 above)
Comment (deepaksrma) [reply os/net/lwip/src/api/sockets.c:393]: done

File: os/include/tinyara/net/net.h (hunk 16 above)
Comment (deepaksrma) [reply os/include/tinyara/net/net.h]: done

File: os/net/socket/net_vfcntl.c [hunk 21]
```diff
@@ -99,7 +99,7 @@
 int net_vfcntl(int sockfd, int cmd, va_list ap)
 {
 
-	FAR struct socket *sock = (struct socket *)get_socket(sockfd);
+	FAR struct socket *sock = (struct socket *)get_socket(sockfd, getpid());
```
Comment (pillip8282) [inline os/net/socket/net_vfcntl.c:102]: I think you can get pid inside get_socket()

File: build/configs/artik053/netmgr/defconfig [hunk 22]
```diff
@@ -656,7 +656,7 @@ CONFIG_NET_IPv6_PMTU_FOR_MULTICAST=y
 # Socket support
 #
 CONFIG_NET_SOCKET=y
-CONFIG_NBSDSOCKET_DESCRIPTORS=8
+CONFIG_NBSDSOCKET_DESCRIPTORS=16
```
Comment (pillip8282) [inline build/configs/artik053/netmgr/defconfig:659]: why did you change value?

File: os/net/lwip/src/include/lwip/sockets.h [hunk 23]
```diff
@@ -157,6 +157,8 @@ struct lwip_setgetsockopt_data {
 	err_t err;
 	/** semaphore to wake up the calling task */
 	void *completed_sem;
+	/** pid of current task */
+	pid_t pid;
```
Comment (pillip8282) [inline os/net/lwip/src/include/lwip/sockets.h:161]: why did you add this?

File: os/net/netmgr/net_vfs.c [hunk 24]
```diff
@@ -178,8 +178,8 @@ int net_ioctl(int sd, int cmd, unsigned long arg)
 	}
 
 	/* ToDo:  Verify that the sd corresponds to valid, allocated socket */
-	sock = get_socket(sd);
-	if (NULL == sock) {
+	sock = get_socket(sd, getpid());
+	if (sock == NULL) {
```
Comment (pillip8282) [inline os/net/netmgr/net_vfs.c:182]: why did you change order?

File: os/net/netmgr/netstack_lwip.c [hunk 25]
```diff
@@ -427,22 +427,24 @@ static void _get_raw_info(struct netmon_sock *sock_info, struct lwip_sock *lsock
 	sock_info->remote.ip.sin_port = 0;
 }
 
-extern struct lwip_sock *get_lwip_sock_info(void);
+extern struct socketlist *get_lwip_sock_info(void);
```
Comment (pillip8282) [inline os/net/netmgr/netstack_lwip.c:430]: socketlist is for TizenRT, lwip_sock is for lwip. get_lwip_sock_info() is implemented in sockets.c which lwip contains. so you shouldn't use socketlist in lwip code. it violates layered dependency.
you'd better to call sched_getsockets() which TizenRT already provides instead of get_lwip_sock_info

**Review by pillip8282 — CHANGES_REQUESTED** [review]
you have to remove sockets[NUM_SOCKETS] which is defined in sockets.c.
And please describe how did you verify the PR.

File: os/net/socket/net_vfcntl.c (hunk 21 above)
Comment (deepaksrma) [reply os/net/socket/net_vfcntl.c:102]: yeah we can get pid inside the get_socket(), but the thing is many time process gets changed. like in event_callback(), get_socket() is called and at that time context switch was happening so could not fetch the correct socketlist this is why i am passing the pid from the function itself. for event_callback() i am passing the pid after taking it from netconn.

File: build/configs/artik053/netmgr/defconfig (hunk 22 above)
Comment (deepaksrma) [reply build/configs/artik053/netmgr/defconfig:659]: because at artik053 when i tried to join the wifi then alloc_socket() was called 9 times so for last socket it was returning -1.

File: os/net/lwip/src/include/lwip/sockets.h (hunk 23 above)
Comment (deepaksrma) [reply os/net/lwip/src/include/lwip/sockets.h:161]: when lwip_setsockopt() is called then it also calls lwip_setsockopt_impl() and for lwip_setsockopt_impl() pid was changed. so i am storing the pid in data in lwip_setsockopt().

File: os/net/netmgr/net_vfs.c (hunk 24 above)
Comment (deepaksrma) [reply os/net/netmgr/net_vfs.c:182]: because i think it looks better than earlier one.

**Comment by deepaksrma** [comment]
> you have to remove sockets[NUM_SOCKETS] which is defined in sockets.c.
- already removed. did you any?

> And please describe how did you verify the PR.
- first follow these steps for reproducing the issue:
Socket bind issue reproducing steps:
Repeat joining the Wi-Fi, and making faults, it will not be recovered after some repeats

Joining steps in Tash
1. wm_test start
2. wm_test join [SSID] wpa2_aes [PASSWORD]

you can change codes in any tash command and call like below.
binary_manager_update_binary("wifi");


and got to know we were getting error like below at 15th repetition:
TASH>>[WM]T150  _wifi_utils_callback_handler:167
run utils callback handler (should be moved to booting)
[WM][ERR] src/wifi_manager/wifi_manager_lwnl_listener.c: 179 line err(File table overflow)
Invalid Scenario.


but with this patch we are not getting any error even after trying for 100 times.
ping is also working.

File: os/net/netmgr/netstack_lwip.c (hunk 25 above)
Comment (deepaksrma) [reply os/net/netmgr/netstack_lwip.c:430]: kept in mind. will change it.

File: os/net/socket/net_vfcntl.c (hunk 21 above)
Comment (pillip8282) [reply os/net/socket/net_vfcntl.c:102]: I can't see the case you told "and at that time context switch was happening so could not fetch the correct socketlist this is why i am passing the pid from the function itself. for event_callback() i am passing the pid after taking it from netconn".
Did you upload it in different PR?

**Comment by pillip8282** [comment]
@deepaksrma 

You'd better to describe how did you test it in commit message. I'll helpful.
> already removed. did you any?
- I can't see - expression(remove sockets) in this PR.

File: os/net/socket/net_vfcntl.c (hunk 21 above)
Comment (deepaksrma) [reply os/net/socket/net_vfcntl.c:102]: pid was getting changed either in event_callback() or lwip_setsockopt_impl() or in both. don't remember exactly. but i am sure in at least one of the mentioned function, pid was getting changed. this is why i had to pass pid as argument to fetch correct socketlist.

no, i didn't upload it in the PR of my forked repo that's why i was getting error.

File: os/net/lwip/src/api/sockets.c [hunk 26]
```diff
@@ -299,13 +299,11 @@ struct lwip_socket_multicast_pair {
...
-static int lwip_socket_register_membership(int s, const ip4_addr_t *if_addr, const ip4_addr_t *multi_addr);
-static void lwip_socket_unregister_membership(int s, const ip4_addr_t *if_addr, const ip4_addr_t *multi_addr);
-static void lwip_socket_drop_registered_memberships(int s);
+static int lwip_socket_register_membership(int s, const ip4_addr_t *if_addr, const ip4_addr_t *multi_addr, pid_t pid);
+static void lwip_socket_unregister_membership(int s, const ip4_addr_t *if_addr, const ip4_addr_t *multi_addr, pid_t pid);
+static void lwip_socket_drop_registered_memberships(struct lwip_sock *sock);
 #endif							/* LWIP_IGMP */
 
-/** The global array of available sockets */
-static struct lwip_sock sockets[NUM_SOCKETS];
```
Comment (deepaksrma) [inline os/net/lwip/src/api/sockets.c:308]: @pillip8282 please see here. 
i have removed sockets array.

**Comment by deepaksrma** [comment]
> @deepaksrma
> 
> You'd better to describe how did you test it in commit message. I'll helpful.
> done.

File: os/net/netmgr/netstack_lwip.c (hunk 25 above)
Comment (deepaksrma) [reply os/net/netmgr/netstack_lwip.c]: done.

**Review by sunghan-chang — APPROVED** [review]

File: os/net/lwip/src/api/sockets.c [hunk 27]
```diff
@@ -2055,6 +2075,7 @@ int lwip_getsockopt(int s, int level, int optname, void *optval, socklen_t *optl
 	LWIP_SETGETSOCKOPT_DATA_VAR_REF(data).level = level;
 	LWIP_SETGETSOCKOPT_DATA_VAR_REF(data).optname = optname;
 	LWIP_SETGETSOCKOPT_DATA_VAR_REF(data).optlen = *optlen;
+	LWIP_SETGETSOCKOPT_DATA_VAR_REF(data).pid = getpid();
```
Comment (pillip8282) [inline os/net/lwip/src/api/sockets.c:2078]: you have to consider following case.
there are Task1(T1) Task2(T2). and they have same socket S1
T1 calls getsockopt, T2 calls setsockopt then calling LWIP_SETGETSOCKOPT_DATA_VAR_REF() can be mixed because of context switching. I think you can solve this problem regarding err_t tcpip_send_msg_wait_sem(tcpip_callback_fn fn, void *apimsg, sys_sem_t *sem); Check op_sync

**Review by pillip8282 — CHANGES_REQUESTED** [review]

File: os/net/lwip/src/api/sockets.c [hunk 28]
```diff
@@ -1319,7 +1342,7 @@ int lwip_writev(int s, const struct iovec *iov, int iovcnt)
  * @param exceptset_out: set os sockets that had error events
  * @return number of sockets that had events (read/write/exception) (>= 0)
  */
-static int lwip_selscan(int maxfdp1, fd_set *readset_in, fd_set *writeset_in, fd_set *exceptset_in, fd_set *readset_out, fd_set *writeset_out, fd_set *exceptset_out)
+static int lwip_selscan(int maxfdp1, fd_set *readset_in, fd_set *writeset_in, fd_set *exceptset_in, fd_set *readset_out, fd_set *writeset_out, fd_set *exceptset_out, pid_t pid)
```
[the remark this answers is not in the document]
Comment (deepaksrma) [reply os/net/lwip/src/api/sockets.c:1345]: done

File: os/net/lwip/src/api/sockets.c [hunk 29]
```diff
@@ -498,6 +511,7 @@ void free_socket(struct lwip_sock *sock, int is_tcp)
 
 	/* Protect socket array */
 	SYS_ARCH_SET(sock->conn, NULL);
+	sched_kfree(sock);
```
[the remark this answers is not in the document]
Comment (deepaksrma) [reply os/net/lwip/src/api/sockets.c:514]: i used kmm_zalloc for lwip_sock allocation so what should i use to free the memory?

File: os/net/lwip/src/api/sockets.c [hunk 30]
```diff
@@ -2970,9 +2990,8 @@ static void lwip_socket_unregister_membership(int s, const ip4_addr_t *if_addr,
  *
  * ATTENTION: this function is NOT called from tcpip_thread (or under CORE_LOCK).
  */
-static void lwip_socket_drop_registered_memberships(int s)
+static void lwip_socket_drop_registered_memberships(struct lwip_sock *sock)
```
[the remark this answers is not in the document]
Comment (deepaksrma) [reply os/net/lwip/src/api/sockets.c:2993]: when we once fetched the sock then what's the need to fetch sock again and again

File: os/net/lwip/src/api/sockets.c [hunk 31]
```diff
@@ -681,6 +690,19 @@ int lwip_sock_close(struct lwip_sock *sock)
...
+
+	DEBUGASSERT(sock->conn);
+	tcb = sched_gettcb(sock->conn->pid);
+	DEBUGASSERT(tcb && tcb->group);
+	list = tcb->group->tg_socketlist;
+	sd = sock->conn->socket - LWIP_SOCKET_OFFSET;
+#if LWIP_IGMP
+	/* drop all possibly joined IGMP memberships */
+	lwip_socket_drop_registered_memberships(sock);
+#endif                                                  /* LWIP_IGMP */
```
[the remark this answers is not in the document]
Comment (deepaksrma) [reply os/net/lwip/src/api/sockets.c:705]: because net_releaselist()  calls lwip_sock_close(). so if i don't call it here then this function will not be called.

File: os/net/lwip/src/api/sockets.c [hunk 32]
```diff
@@ -658,19 +671,15 @@ int lwip_close(int s)
 	struct lwip_sock *sock;
 
 	LWIP_DEBUGF(SOCKETS_DEBUG, ("lwip_close(%d)\n", s));
+	sock = get_socket(s, getpid());
```
[the remark this answers is not in the document]
Comment (deepaksrma) [reply os/net/lwip/src/api/sockets.c:677]: modified. please check

File: os/net/lwip/src/api/sockets.c (hunk 27 above)
Comment (deepaksrma) [reply os/net/lwip/src/api/sockets.c:2073]: can you please clear it more "I think you can solve this problem regarding err_t tcpip_send_msg_wait_sem(tcpip_callback_fn fn, void *apimsg, sys_sem_t *sem); Check op_sync" ??

File: os/net/socket/net_sockets.c [hunk 33]
```diff
@@ -133,10 +136,36 @@ void net_initlist(FAR struct socketlist *list)
...
+		if (list->sl_sockets[i].sock) {
+
+			psock = (struct lwip_sock *)list->sl_sockets[i].sock;
+
+			if (list->sl_sockets[i].s_crefs == 1) {
+				ret = lwip_sock_close(psock);
+				if (ret) {
+					ndbg("socket could not close properly\n");
+					list->sl_sockets[i].sock = NULL;
+					(void)sem_destroy(&list->sl_sem);
```
Comment (amogh-samsung) [inline os/net/socket/net_sockets.c:153]: You could replace this with a "break" statement to avoid duplication?

Comment (deepaksrma) [reply os/net/socket/net_sockets.c:153]: can you please clear it more?
where is duplication?

**Comment by sunghan-chang** [comment]
@deepaksrma Please fix violations of coding rules.
```
Target : [feb4463] - Code Rule Check Result:
os/net/netmgr/netstack_lwip.c:509: ERROR: [BRC_M_SMT] else should follow close brace '}'
os/net/socket/net_sockets.c:161: ERROR: [BRC_M_SMT] else should follow close brace '}'
```

**Comment by sunghan-chang** [comment]
@deepaksrma Could you let us know a test result with current code?

**Comment by deepaksrma** [comment]
> @deepaksrma Could you let us know a test result with current code?

ping is working on artik053 and mtk board.
issue is checked 100 times at mtk board, no error found.

could not test on rtk because it has some TZ memory issue.

File: os/net/lwip/src/api/sockets.c (hunk 29 above)
[the remark this answers is not in the document]
Comment (pillip8282) [reply os/net/lwip/src/api/sockets.c:515]: You can call kmm_free()

File: os/net/lwip/src/api/sockets.c (hunk 27 above)
Comment (pillip8282) [reply os/net/lwip/src/api/sockets.c:2075]: for example T1, T2 use same netconn because they access same socket, lwIP sends signal to both tasks with same netconn->op_complete.  To prevent this op_sync is added.
I think same things could be heppend to lwip_setsockopt too.
Or you can consider LWIP_NETCONN_SEM_PER_THREAD option which was not defined when op_sync is added.
I found it recently.

File: os/net/lwip/src/api/sockets.c (hunk 29 above)
[the remark this answers is not in the document]
Comment (deepaksrma) [reply os/net/lwip/src/api/sockets.c:515]: ok

**Review by pillip8282 — APPROVED** [review]
lgtm
